File: lambda/user_journey/validation.go
package user_journey

import (
	"encoding/json"
	"fmt"
	"net/url"
	"regexp"
	"strings"
	"unicode"
)

// ImportSchema represents the expected structure for import_usecase
type ImportSchema struct {
	ExportVersion string           `json:"exportVersion"`
	ExportedAt    string           `json:"exportedAt"`
	Usecase       UsecaseSchema    `json:"usecase"`
	Steps         []StepSchema     `json:"steps"`
	Variables     []VariableSchema `json:"variables"`
	Secrets       []SecretSchema   `json:"secrets"`
	Hooks         *HooksSchema     `json:"hooks,omitempty"`
}

type UsecaseSchema struct {
	Name        string   `json:"name"`
	Description string   `json:"description"`
	StartingURL string   `json:"starting_url"`
	Active      bool     `json:"active"`
	Tags        []string `json:"tags"`
}

type StepSchema struct {
	Sort               int    `json:"sort"`
	Instruction        string `json:"instruction"`
	StepType           string `json:"step_type"`
	SecretKey          string `json:"secret_key,omitempty"`
	CaptureVariable    string `json:"capture_variable,omitempty"`
	ValidationType     string `json:"validation_type,omitempty"`
	ValidationOperator string `json:"validation_operator,omitempty"`
	ValidationValue    string `json:"validation_value,omitempty"`
	AssertionVariable  string `json:"assertion_variable,omitempty"`
	ValueStep          string `json:"value_step,omitempty"`
	ValueType          string `json:"value_type,omitempty"`
}

type VariableSchema struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

type SecretSchema struct {
	Key         string  `json:"key"`
	Description string  `json:"description"`
	Value       *string `json:"value"`
	Placeholder string  `json:"placeholder"`
}

type HooksSchema struct {
	BeforeScript string `json:"beforeScript"`
	AfterScript  string `json:"afterScript"`
}

// ValidationResult represents the result of JSON validation
type ValidationResult struct {
	IsValid bool          `json:"isValid"`
	Errors  []string      `json:"errors"`
	Schema  *ImportSchema `json:"schema,omitempty"`
}

// ValidateAndParseJSON validates the generated JSON against the import_usecase schema
func ValidateAndParseJSON(jsonStr string) *ValidationResult {
	result := &ValidationResult{
		IsValid: false,
		Errors:  []string{},
	}

	// First, check if it's valid JSON
	var rawData interface{}
	if err := json.Unmarshal([]byte(jsonStr), &rawData); err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("Invalid JSON format: %v", err))
		return result
	}

	// Try to parse into our schema
	var schema ImportSchema
	if err := json.Unmarshal([]byte(jsonStr), &schema); err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("JSON does not match expected schema: %v", err))
		return result
	}

	// Validate required fields and structure
	errors := validateImportSchema(&schema)
	if len(errors) > 0 {
		result.Errors = errors
		return result
	}

	result.IsValid = true
	result.Schema = &schema
	return result
}

// validateImportSchema performs detailed validation of the import schema
func validateImportSchema(schema *ImportSchema) []string {
	var errors []string

	// Validate export version
	if schema.ExportVersion == "" {
		errors = append(errors, "exportVersion is required")
	} else if schema.ExportVersion != "1.0" {
		errors = append(errors, "exportVersion must be '1.0'")
	}

	// Validate exportedAt
	if schema.ExportedAt == "" {
		errors = append(errors, "exportedAt is required")
	}

	// Validate usecase
	usecaseErrors := validateUsecaseSchema(&schema.Usecase)
	errors = append(errors, usecaseErrors...)

	// Validate steps
	if len(schema.Steps) == 0 {
		errors = append(errors, "at least one step is required")
	} else {
		stepErrors := validateStepsSchema(schema.Steps)
		errors = append(errors, stepErrors...)
	}

	// Validate variables (optional but if present, must be valid)
	if schema.Variables != nil {
		variableErrors := validateVariablesSchema(schema.Variables)
		errors = append(errors, variableErrors...)
	}

	// Validate secrets (optional but if present, must be valid)
	if schema.Secrets != nil {
		secretErrors := validateSecretsSchema(schema.Secrets)
		errors = append(errors, secretErrors...)
	}

	return errors
}

// validateUsecaseSchema validates the usecase object
func validateUsecaseSchema(usecase *UsecaseSchema) []string {
	var errors []string

	if strings.TrimSpace(usecase.Name) == "" {
		errors = append(errors, "usecase.name is required")
	}

	if strings.TrimSpace(usecase.Description) == "" {
		errors = append(errors, "usecase.description is required")
	}

	if strings.TrimSpace(usecase.StartingURL) == "" {
		errors = append(errors, "usecase.starting_url is required")
	} else if !strings.HasPrefix(usecase.StartingURL, "http://") && !strings.HasPrefix(usecase.StartingURL, "https://") {
		errors = append(errors, "usecase.starting_url must be a valid HTTP/HTTPS URL")
	}

	if usecase.Tags == nil {
		errors = append(errors, "usecase.tags must be an array (can be empty)")
	}

	return errors
}

// validateStepsSchema validates the steps array
func validateStepsSchema(steps []StepSchema) []string {
	var errors []string
	validStepTypes := map[string]bool{
		"navigation":     true,
		"validation":     true,
		"secret":         true,
		"retrieve_value": true,
		"assertion":      true,
	}
	validOperators := map[string]bool{
		"exact":                     true,
		"exact_case_insensitive":    true,
		"contains":                  true,
		"contains_case_insensitive": true,
		"equals":                    true,
		"not_equals":                true,
		"greater_then":              true,
		"less_then":                 true,
		"greater_or_equal_then":     true,
		"less_or_equal_then":        true,
	}

	sortNumbers := make(map[int]bool)

	for i, step := range steps {
		stepPrefix := fmt.Sprintf("steps[%d]", i)

		// Validate sort number
		if step.Sort <= 0 {
			errors = append(errors, fmt.Sprintf("%s.sort must be a positive integer", stepPrefix))
		}
		if sortNumbers[step.Sort] {
			errors = append(errors, fmt.Sprintf("%s.sort %d is duplicated", stepPrefix, step.Sort))
		}
		sortNumbers[step.Sort] = true

		// Validate instruction
		if strings.TrimSpace(step.Instruction) == "" {
			errors = append(errors, fmt.Sprintf("%s.instruction is required", stepPrefix))
		}

		// Validate step_type
		if !validStepTypes[step.StepType] {
			errors = append(errors, fmt.Sprintf("%s.step_type must be one of: navigation, validation, secret, retrieve_value", stepPrefix))
		}

		// Validate validation-specific fields
		if step.StepType == "validation" {
			if step.ValidationType == "" {
				errors = append(errors, fmt.Sprintf("%s.validation_type is required for validation steps", stepPrefix))
			}
			if step.ValidationOperator != "" && !validOperators[step.ValidationOperator] {
				errors = append(errors, fmt.Sprintf("%s.validation_operator must be one of: exact, exact_case_insensitive, contains, contains_case_insensitive, equals, greater_then, less_then, greater_or_equal_then, less_or_equal_then", stepPrefix))
			}
		}

		// Validate secret-specific fields
		if step.StepType == "secret" && strings.TrimSpace(step.SecretKey) == "" {
			errors = append(errors, fmt.Sprintf("%s.secret_key is required for secret steps", stepPrefix))
		}

		// Validate retrieve_value-specific fields
		if step.StepType == "retrieve_value" {
			if strings.TrimSpace(step.CaptureVariable) == "" {
				errors = append(errors, fmt.Sprintf("%s.capture_variable is required for retrieve_value steps", stepPrefix))
			}
			if strings.TrimSpace(step.ValueType) == "" {
				errors = append(errors, fmt.Sprintf("%s.value_type is required for retrieve_value steps", stepPrefix))
			}
		}
	}

	return errors
}

// validateVariablesSchema validates the variables array
func validateVariablesSchema(variables []VariableSchema) []string {
	var errors []string
	keys := make(map[string]bool)

	for i, variable := range variables {
		varPrefix := fmt.Sprintf("variables[%d]", i)

		if strings.TrimSpace(variable.Key) == "" {
			errors = append(errors, fmt.Sprintf("%s.key is required", varPrefix))
		} else if keys[variable.Key] {
			errors = append(errors, fmt.Sprintf("%s.key '%s' is duplicated", varPrefix, variable.Key))
		}
		keys[variable.Key] = true

		if strings.TrimSpace(variable.Value) == "" {
			errors = append(errors, fmt.Sprintf("%s.value is required", varPrefix))
		}
	}

	return errors
}

// validateSecretsSchema validates the secrets array
func validateSecretsSchema(secrets []SecretSchema) []string {
	var errors []string
	keys := make(map[string]bool)

	for i, secret := range secrets {
		secretPrefix := fmt.Sprintf("secrets[%d]", i)

		if strings.TrimSpace(secret.Key) == "" {
			errors = append(errors, fmt.Sprintf("%s.key is required", secretPrefix))
		} else if keys[secret.Key] {
			errors = append(errors, fmt.Sprintf("%s.key '%s' is duplicated", secretPrefix, secret.Key))
		}
		keys[secret.Key] = true

		if strings.TrimSpace(secret.Description) == "" {
			errors = append(errors, fmt.Sprintf("%s.description is required", secretPrefix))
		}
	}

	return errors
}

// Security patterns for input validation
var (
	// XSS detection patterns
	xssPattern = regexp.MustCompile(`(?i)<script|javascript:|on\w+\s*=|<iframe|<object|<embed`)

	// Path traversal detection patterns
	pathTraversalPattern = regexp.MustCompile(`\.\.[\/\\]|%2e%2e[\/\\]|\.\.%2f|\.\.%5c`)

	// Valid title pattern (alphanumeric, spaces, hyphens, underscores)
	titlePattern = regexp.MustCompile(`^[a-zA-Z0-9\s\-_]+$`)
)

// SecurityValidationResult represents the result of security validation
type SecurityValidationResult struct {
	IsSecure bool     `json:"isSecure"`
	Threats  []string `json:"threats"`
}

// ValidateInputSecurity checks input for security threats
func ValidateInputSecurity(input string) *SecurityValidationResult {
	result := &SecurityValidationResult{
		IsSecure: true,
		Threats:  []string{},
	}

	if xssPattern.MatchString(input) {
		result.IsSecure = false
		result.Threats = append(result.Threats, "Potential XSS content detected")
	}

	if pathTraversalPattern.MatchString(input) {
		result.IsSecure = false
		result.Threats = append(result.Threats, "Potential path traversal content detected")
	}

	return result
}

// SanitizeInput removes potentially dangerous characters and normalizes input
func SanitizeInput(input string) string {
	// Remove control characters except newlines and tabs
	sanitized := strings.Map(func(r rune) rune {
		if unicode.IsControl(r) && r != '\n' && r != '\t' && r != '\r' {
			return -1
		}
		return r
	}, input)

	// Normalize whitespace
	sanitized = regexp.MustCompile(`\s+`).ReplaceAllString(sanitized, " ")

	// Trim whitespace
	sanitized = strings.TrimSpace(sanitized)

	return sanitized
}

// ValidateTitle validates and sanitizes a title input
func ValidateTitle(title string) (string, []string) {
	var errors []string

	// Sanitize input
	sanitized := SanitizeInput(title)

	// Security check
	securityResult := ValidateInputSecurity(sanitized)
	if !securityResult.IsSecure {
		errors = append(errors, securityResult.Threats...)
		return "", errors
	}

	// Length validation
	if len(sanitized) == 0 {
		errors = append(errors, "Title is required")
		return "", errors
	}

	if len(sanitized) > 200 {
		errors = append(errors, "Title must be 200 characters or less")
		return "", errors
	}

	if len(sanitized) < 3 {
		errors = append(errors, "Title must be at least 3 characters long")
		return "", errors
	}

	// Pattern validation
	if !titlePattern.MatchString(sanitized) {
		errors = append(errors, "Title can only contain letters, numbers, spaces, hyphens, and underscores")
		return "", errors
	}

	// Check for meaningful content
	words := strings.Fields(sanitized)
	if len(words) == 0 {
		errors = append(errors, "Title must contain meaningful content")
		return "", errors
	}

	return sanitized, errors
}

// ValidateURL validates and sanitizes a URL input
func ValidateURL(urlStr string) (string, []string) {
	var errors []string

	// Sanitize input
	sanitized := SanitizeInput(urlStr)

	// Security check
	securityResult := ValidateInputSecurity(sanitized)
	if !securityResult.IsSecure {
		errors = append(errors, securityResult.Threats...)
		return "", errors
	}

	// Required validation
	if len(sanitized) == 0 {
		errors = append(errors, "Starting URL is required")
		return "", errors
	}

	// URL format validation
	parsedURL, err := url.Parse(sanitized)
	if err != nil {
		errors = append(errors, "Invalid URL format")
		return "", errors
	}

	// Scheme validation
	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
		errors = append(errors, "URL must start with http:// or https://")
		return "", errors
	}

	// Host validation
	if parsedURL.Host == "" {
		errors = append(errors, "URL must include a valid host")
		return "", errors
	}

	// Check for suspicious hosts
	if strings.Contains(parsedURL.Host, "..") || len(parsedURL.Host) < 3 {
		errors = append(errors, "URL host appears to be invalid")
		return "", errors
	}

	return sanitized, errors
}

// ValidateUserJourney validates and sanitizes a user journey description
func ValidateUserJourney(journey string) (string, []string) {
	var errors []string

	// Sanitize input
	sanitized := SanitizeInput(journey)

	// Security check
	securityResult := ValidateInputSecurity(sanitized)
	if !securityResult.IsSecure {
		errors = append(errors, securityResult.Threats...)
		return "", errors
	}

	// Required validation
	if len(sanitized) == 0 {
		errors = append(errors, "User journey description is required")
		return "", errors
	}

	// Length validation
	if len(sanitized) < 50 {
		errors = append(errors, "User journey description must be at least 50 characters")
		return "", errors
	}

	if len(sanitized) > 2000 {
		errors = append(errors, "User journey description must be 2000 characters or less")
		return "", errors
	}

	// Content validation
	words := strings.Fields(sanitized)
	if len(words) < 10 {
		errors = append(errors, "User journey description should contain more detailed steps")
		return "", errors
	}

	// Check for action words
	actionWords := []string{"click", "enter", "navigate", "select", "submit", "verify", "check", "fill", "choose", "confirm", "type", "press", "scroll", "hover"}
	hasActions := false
	lowerJourney := strings.ToLower(sanitized)

	for _, action := range actionWords {
		if strings.Contains(lowerJourney, action) {
			hasActions = true
			break
		}
	}

	if !hasActions {
		errors = append(errors, "User journey should include specific actions (click, enter, navigate, etc.)")
		return "", errors
	}

	return sanitized, errors
}

// SanitizeAndFixJSON attempts to fix common issues in generated JSON
func SanitizeAndFixJSON(jsonStr string) string {
	// Remove any markdown code block markers
	jsonStr = strings.ReplaceAll(jsonStr, "```json", "")
	jsonStr = strings.ReplaceAll(jsonStr, "```", "")

	// Trim whitespace
	jsonStr = strings.TrimSpace(jsonStr)

	// Ensure it starts and ends with braces
	if !strings.HasPrefix(jsonStr, "{") {
		start := strings.Index(jsonStr, "{")
		if start != -1 {
			jsonStr = jsonStr[start:]
		}
	}

	if !strings.HasSuffix(jsonStr, "}") {
		end := strings.LastIndex(jsonStr, "}")
		if end != -1 {
			jsonStr = jsonStr[:end+1]
		}
	}

	return jsonStr
}

File: web-app/frontend/src/utils/validation.ts
import { errorManager } from './errorManager';

export interface ValidationRule {
  required?: boolean;
  minLength?: number;
  maxLength?: number;
  pattern?: RegExp;
  custom?: (value: string) => string | null;
}

export interface ValidationResult {
  isValid: boolean;
  error?: string;
  warning?: string;
  suggestion?: string;
}

export interface FieldValidationConfig {
  rules: ValidationRule;
  label: string;
  description?: string;
  examples?: string[];
  hints?: string[];
}

// Security patterns for input sanitization
const SECURITY_PATTERNS = {
  // Detect potential XSS attempts
  xss: /<script|javascript:|on\w+\s*=|<iframe|<object|<embed/i,
  // Detect path traversal attempts
  pathTraversal: /\.\.[\/\\]|%2e%2e[\/\\]|\.\.%2f|\.\.%5c/i,
};

export class ValidationManager {
  private static instance: ValidationManager;

  static getInstance(): ValidationManager {
    if (!ValidationManager.instance) {
      ValidationManager.instance = new ValidationManager();
    }
    return ValidationManager.instance;
  }

  // Sanitize input to prevent security issues (light sanitization for real-time input)
  sanitizeInput(value: string): string {
    if (!value) return value;

    // Only remove the most dangerous XSS characters during typing
    let sanitized = value
      .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
      .replace(/javascript:/gi, '')
      .replace(/on\w+\s*=/gi, '');

    return sanitized;
  }

  // Full sanitization for final validation and submission
  sanitizeForValidation(value: string): string {
    if (!value) return value;

    // Remove potential XSS characters
    let sanitized = value
      .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
      .replace(/javascript:/gi, '')
      .replace(/on\w+\s*=/gi, '');

    // Normalize whitespace only for validation
    sanitized = sanitized.replace(/\s+/g, ' ').trim();

    return sanitized;
  }

  // Check for security threats in input
  checkSecurity(value: string): ValidationResult {
    if (!value) return { isValid: true };

    for (const [type, pattern] of Object.entries(SECURITY_PATTERNS)) {
      if (pattern.test(value)) {
        return {
          isValid: false,
          error: `Input contains potentially unsafe content (${type}). Please remove special characters and try again.`
        };
      }
    }

    return { isValid: true };
  }

  // Validate individual field
  validateField(value: string, config: FieldValidationConfig): ValidationResult {
    const { rules, label } = config;

    // Security check first
    const securityResult = this.checkSecurity(value);
    if (!securityResult.isValid) {
      return securityResult;
    }

    // Sanitize input for validation
    const sanitizedValue = this.sanitizeForValidation(value);

    // Required validation
    if (rules.required && !sanitizedValue.trim()) {
      return {
        isValid: false,
        error: `${label} is required`,
        suggestion: config.examples?.[0] ? `Example: ${config.examples[0]}` : undefined
      };
    }

    // Skip other validations if field is empty and not required
    if (!sanitizedValue.trim() && !rules.required) {
      return { isValid: true };
    }

    // Length validations
    if (rules.minLength && sanitizedValue.length < rules.minLength) {
      return {
        isValid: false,
        error: `${label} must be at least ${rules.minLength} characters`,
        suggestion: `Current length: ${sanitizedValue.length}. Please add more detail.`
      };
    }

    if (rules.maxLength && sanitizedValue.length > rules.maxLength) {
      return {
        isValid: false,
        error: `${label} must be ${rules.maxLength} characters or less`,
        suggestion: `Current length: ${sanitizedValue.length}. Please shorten your input.`
      };
    }

    // Pattern validation
    if (rules.pattern && !rules.pattern.test(sanitizedValue)) {
      return this.getPatternError(rules.pattern, label, config.examples);
    }

    // Custom validation
    if (rules.custom) {
      const customError = rules.custom(sanitizedValue);
      if (customError) {
        return {
          isValid: false,
          error: customError
        };
      }
    }

    // Check for warnings (non-blocking issues)
    const warning = this.getFieldWarning(sanitizedValue, config);

    return {
      isValid: true,
      warning: warning || undefined
    };
  }

  private getPatternError(pattern: RegExp, label: string, examples?: string[]): ValidationResult {
    // URL pattern
    if (pattern.source.includes('https?')) {
      return {
        isValid: false,
        error: 'Please enter a valid URL starting with http:// or https://',
        suggestion: examples?.[0] || 'Example: https://example.com'
      };
    }

    // Title pattern (alphanumeric with spaces, hyphens, underscores)
    if (pattern.source.includes('a-zA-Z0-9')) {
      return {
        isValid: false,
        error: `${label} can only contain letters, numbers, spaces, hyphens, and underscores`,
        suggestion: 'Avoid special characters like @, #, $, %, etc.'
      };
    }

    return {
      isValid: false,
      error: `${label} format is invalid`
    };
  }

  private getFieldWarning(value: string, config: FieldValidationConfig): string | null {
    // Check for very short descriptions that might not be detailed enough
    if (config.label.toLowerCase().includes('description') && value.length < 100) {
      return 'Consider adding more detail for better AI generation results';
    }

    // Check for URLs without HTTPS
    if (config.label.toLowerCase().includes('url') && value.startsWith('http://')) {
      return 'Consider using HTTPS for better security';
    }

    // Check for overly simple titles
    if (config.label.toLowerCase().includes('title') && value.split(' ').length < 3) {
      return 'More descriptive titles help organize your test cases';
    }

    return null;
  }

  // Get helpful hints for a field
  getFieldHints(config: FieldValidationConfig, currentValue: string = ''): string[] {
    const hints: string[] = [...(config.hints || [])];

    // Dynamic hints based on current state
    if (config.rules.maxLength) {
      const remaining = config.rules.maxLength - currentValue.length;
      if (remaining < 50 && remaining > 0) {
        hints.push(`${remaining} characters remaining`);
      }
    }

    if (config.rules.minLength && currentValue.length < config.rules.minLength) {
      const needed = config.rules.minLength - currentValue.length;
      hints.push(`${needed} more characters needed`);
    }

    return hints;
  }

  // Validate entire form
  validateForm(formData: Record<string, string>, configs: Record<string, FieldValidationConfig>): {
    isValid: boolean;
    errors: Record<string, string>;
    warnings: Record<string, string>;
  } {
    const errors: Record<string, string> = {};
    const warnings: Record<string, string> = {};

    for (const [field, config] of Object.entries(configs)) {
      const value = formData[field] || '';
      const result = this.validateField(value, config);

      if (!result.isValid && result.error) {
        errors[field] = result.error;
      }

      if (result.warning) {
        warnings[field] = result.warning;
      }
    }

    return {
      isValid: Object.keys(errors).length === 0,
      errors,
      warnings
    };
  }
}

// Field configurations for User Journey Wizard
export const WIZARD_FIELD_CONFIGS: Record<string, FieldValidationConfig> = {
  title: {
    rules: {
      required: true,
      maxLength: 200,
      pattern: /^[a-zA-Z0-9\s\-_]+$/,
      custom: (value: string) => {
        // Check for meaningful content
        if (value.trim().length < 3) {
          return 'Title should be at least 3 characters long';
        }
        
        // Check for excessive repetition
        const words = value.toLowerCase().split(/\s+/);
        const uniqueWords = new Set(words);
        if (words.length > 3 && uniqueWords.size < words.length * 0.5) {
          return 'Title contains too much repetition';
        }
        
        return null;
      }
    },
    label: 'Use Case Title',
    description: 'A descriptive name for your test case',
    examples: [
      'User Login Flow Test',
      'E-commerce Checkout Process',
      'Contact Form Submission',
      'Product Search and Filter'
    ],
    hints: [
      'Use descriptive names that clearly identify the test purpose',
      'Avoid special characters except hyphens and underscores',
      'Keep it concise but informative'
    ]
  },
  startingUrl: {
    rules: {
      required: true,
      pattern: /^https?:\/\/.+/,
      custom: (value: string) => {
        try {
          const url = new URL(value);
          
          // Check for localhost or development URLs
          if (url.hostname === 'localhost' || url.hostname.includes('127.0.0.1')) {
            return null; // Allow localhost for development
          }
          
          // Check for suspicious domains
          if (url.hostname.includes('..') || url.hostname.length < 3) {
            return 'URL appears to be invalid or suspicious';
          }
          
          return null;
        } catch {
          return 'Please enter a valid URL';
        }
      }
    },
    label: 'Starting URL',
    description: 'The URL where your user journey begins',
    examples: [
      'https://example.com/login',
      'https://myapp.com/dashboard',
      'https://shop.example.com',
      'https://localhost:3000/app'
    ],
    hints: [
      'Include the full URL with http:// or https://',
      'Make sure the URL is accessible and loads correctly',
      'Use the exact URL where the test should start'
    ]
  },
  userJourney: {
    rules: {
      required: true,
      minLength: 50,
      maxLength: 2000,
      custom: (value: string) => {
        // Check for meaningful content
        const words = value.trim().split(/\s+/);
        if (words.length < 10) {
          return 'User journey should contain more detailed steps';
        }
        
        // Check for action words that indicate a good user journey
        const actionWords = ['click', 'enter', 'navigate', 'select', 'submit', 'verify', 'check', 'fill', 'choose', 'confirm'];
        const hasActions = actionWords.some(word => 
          value.toLowerCase().includes(word)
        );
        
        if (!hasActions) {
          return 'User journey should include specific actions (click, enter, navigate, etc.)';
        }
        
        return null;
      }
    },
    label: 'User Journey Description',
    description: 'Describe the complete user journey step by step',
    examples: [
      'User navigates to login page, enters email and password, clicks login button, and should be redirected to dashboard with welcome message displayed',
      'User searches for a product, filters by price range, adds item to cart, proceeds to checkout, enters shipping information, and completes payment',
      'User fills out contact form with name, email, and message, submits the form, and receives confirmation message'
    ],
    hints: [
      'Include specific actions the user will take',
      'Describe expected outcomes and validations',
      'Mention specific UI elements when possible',
      'Include error scenarios if relevant',
      'Be as detailed as possible for better AI generation'
    ]
  }
};

export const validationManager = ValidationManager.getInstance();